### DISCUSSION TITLE
Add system packages for python updates

### DISCUSSION THREAD

**Discussion by JanMalte** [renovatebot/renovate#14596] [Request Help]
Some python packages rely on additional system packages.
Most common is the `psycopg2` package and its system dependency to the `pg_config` executable.

Using the GitLab renovate-runner self-hosting setup, the `Pipfile.lock` can not be generated, as the PostgreSQL system package is not installed in the renovate python image.

How can I run some pre-setup commands for the renovate_child image used by the renovate runner?

`WARN: artifactErrors (repository=gegl-group/slow-consulting, branch=renovate/psycopg2-2.x)`
```json
{
       "artifactErrors": [
         {
           "lockFile": "Pipfile.lock",
           "stderr": "Command failed: docker run --rm --name=renovate_python --label=renovate_child docker.io/renovate/python:3.8.12 bash -l -c \"pip install --user pipenv && pipenv lock\" …"
         }
       ]
}
```

```
Error: pg_config executable not found.
pg_config is required to build psycopg2 from source.  Please add the directory
containing pg_config to the $PATH or specify the full executable path with the
option:
    python setup.py build_ext --pg-config /path/to/pg_config build ...
```

```
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
wagtail = "~=2.12.2"
gunicorn = "~=20.0.4"
psycopg2 = "~=2.8.6"
django-redis = "~=4.12.1"

[dev-packages]

[requires]
python_version = "3.8"
```

**Comment by rarkins** [comment]
Can you put that into a reproduction repo we can reproduce the problem against?

**Reply by JanMalte** [reply]
There is a closed issue here: https://github.com/renovatebot/renovate/issues/9485

To quote @bagage :
```
# OK
docker run --rm python:3.9.3 pip install psycopg2
# KO, should be working?
docker run --rm docker.io/renovate/python:3.9.3 pip install psycopg2
```

**Reply by JanMalte** [reply]
You already created a draft PR for it: 
https://github.com/containerbase/buildpack/pull/30

**Reply by JanMalte** [reply]
The question is, is there any workaround for now to install the package right before renovate is running?

**Reply by rarkins** [reply]
I'm looking for a poetry repo to reproduce the failure with, so that I have something to test against

**Reply by JanMalte** [reply]
Maybe this might be enough … https://github.com/JanMalte/renovate-pg-config

**Reply by rarkins** [reply]
I tested it and it did not reproduce any error. Did you test it?

**Reply by JanMalte** [reply]
I tried, but the GitHub Renovate App didn't created any onboarding PR right now … is there any way to test if manually?

Running the docker commands from @bagage still results in the error.

**Reply by JanMalte** [reply]
`poetry` is installed using a script called `install-tool` and updating `psycopg2` seems to be working
```
docker run --rm -v $(pwd):/mnt/src/ -w /mnt/src docker.io/renovate/python:3.10.2 bash -l -c "install-tool poetry 1.1.13 && poetry update --lock --no-interaction psycopg2"
```

`pipenv` is installed just with `pip` and fails to update `psycopg2`
```
docker run --rm -v $(pwd):/mnt/src/ -w /mnt/src docker.io/renovate/python:3.10.2 bash -l -c "pip install --user pipenv && pipenv lock"
```

**Reply by JanMalte** [reply]
@rarkins I managed to split the issue into serveral repos

Failing update with `pipenv` (using plain `pip install pipenv` in the docker image):
https://github.com/JanMalte/renovate-psycopg2-pipenv/pull/2

Working update with `poetry` (using the `install_tool` script in the docker image):
https://github.com/JanMalte/renovate-psycopg2-poetry/pull/3

Working update with `pip` (no lock file generation/package install needed):
https://github.com/JanMalte/renovate-psycopg2-pip/pull/2

**Reply by rarkins** [reply]
Strange that `poetry` works while `pipenv` doesn't. I'm not sure it's due to `install-tool` vs `pip install` and instead maybe that `pipenv` builds it from source while `poetry` does not.

**Comment by hugoboos** [comment]
We have the same problem, but with `pip-tools`.

pip-tools can't generate the lock file (requirements.txt) because of: `Error: pg_config executable not found.`

Is there a way to tell Renovate to install the `libpq-dev` system package?

<details>
<summary>Artifact update problem log</summary
Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v "/builds/commonground/core/renovate/renovate/repos/gitlab/commonground/don/developer.overheid.nl":"/builds/commonground/core/renovate/renovate/repos/gitlab/commonground/don/developer.overheid.nl" -v "/builds/commonground/core/renovate/renovate/cache":"/builds/commonground/core/renovate/renovate/cache" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w "/builds/commonground/core/renovate/renovate/repos/gitlab/commonground/don/developer.overheid.nl/api" docker.io/containerbase/sidecar bash -l -c "install-tool python 3.11.2 && pip install --user pip-tools && pip-compile --generate-hashes requirements.in"
WARNING: the legacy dependency resolver is deprecated and will be removed in future versions of pip-tools. The default resolver will be changed to 'backtracking' in pip-tools 7.0.0. Specify --resolver=backtracking to silence this warning.
    error: subprocess-exited-with-error
    
    × python setup.py egg_info did not run successfully.
    │ exit code: 1
    ╰─> [25 lines of output]
        /opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/setuptools/config/setupcfg.py:508: SetuptoolsDeprecationWarning: The license_file parameter is deprecated, use license_files instead.
          warnings.warn(msg, warning_class)
        running egg_info
        creating /tmp/pip-pip-egg-info-414ckik_/psycopg2.egg-info
        writing /tmp/pip-pip-egg-info-414ckik_/psycopg2.egg-info/PKG-INFO
        writing dependency_links to /tmp/pip-pip-egg-info-414ckik_/psycopg2.egg-info/dependency_links.txt
        writing top-level names to /tmp/pip-pip-egg-info-414ckik_/psycopg2.egg-info/top_level.txt
        writing manifest file '/tmp/pip-pip-egg-info-414ckik_/psycopg2.egg-info/SOURCES.txt'
        
        Error: pg_config executable not found.
        
        pg_config is required to build psycopg2 from source.  Please add the directory
        containing pg_config to the $PATH or specify the full executable path with the
        option:
        
            python setup.py build_ext --pg-config /path/to/pg_config build ...
        
        or with the pg_config option in 'setup.cfg'.
        
        If you prefer to avoid building psycopg2 from source, please install the PyPI
        'psycopg2-binary' package instead.
        
        For further information please check the 'doc/src/install.rst' file (also at
        <https://www.psycopg.org/docs/install.html>).
        
        [end of output]
    
    note: This error originates from a subprocess, and is likely not a problem with pip.
Traceback (most recent call last):
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/operations/build/metadata_legacy.py", line 64, in generate_metadata
    call_subprocess(
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/utils/subprocess.py", line 224, in call_subprocess
    raise error
pip._internal.exceptions.InstallationSubprocessError: python setup.py egg_info exited with 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ubuntu/.local/bin/pip-compile", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/piptools/scripts/compile.py", line 589, in cli
    results = resolver.resolve(max_rounds=max_rounds)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/piptools/resolver.py", line 253, in resolve
    has_changed, best_matches = self._resolve_one_round()
                                ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/piptools/resolver.py", line 343, in _resolve_one_round
    their_constraints.extend(self._iter_dependencies(best_match))
  File "/home/ubuntu/.local/lib/python3.11/site-packages/piptools/resolver.py", line 456, in _iter_dependencies
    dependencies = self.repository.get_dependencies(ireq)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/piptools/repositories/local.py", line 87, in get_dependencies
    return self.repository.get_dependencies(ireq)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/piptools/repositories/pypi.py", line 238, in get_dependencies
    self._dependencies_cache[ireq] = self.resolve_reqs(
                                     ^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.local/lib/python3.11/site-packages/piptools/repositories/pypi.py", line 201, in resolve_reqs
    results = resolver._resolve_one(reqset, ireq)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/resolution/legacy/resolver.py", line 509, in _resolve_one
    dist = self._get_dist_for(req_to_install)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/resolution/legacy/resolver.py", line 462, in _get_dist_for
    dist = self.preparer.prepare_linked_requirement(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/operations/prepare.py", line 491, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/operations/prepare.py", line 577, in _prepare_linked_requirement
    dist = _get_prepared_distribution(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/operations/prepare.py", line 69, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/distributions/sdist.py", line 61, in prepare_distribution_metadata
    self.req.prepare_metadata()
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/req/req_install.py", line 545, in prepare_metadata
    self.metadata_directory = generate_metadata_legacy(
                              ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/buildpack/tools/python/3.11.2/lib/python3.11/site-packages/pip/_internal/operations/build/metadata_legacy.py", line 71, in generate_metadata
    raise MetadataGenerationFailed(package_details=details) from error
pip._internal.exceptions.MetadataGenerationFailed: metadata generation failed
</details>

**Reply by viceice** [reply]
- https://github.com/containerbase/base/issues/98
